llvm/lib/Target/Mips/MipsSEFrameLowering.cpp

```diff
@@ -935,21 +935,57 @@ bool MipsSEFrameLowering::assignCalleeSavedSpillSlots(
   if (!STI.hasNanoMips())
     return false;
 
+  std::unordered_map<unsigned, unsigned> Regs{
+      {Mips::GP_NM, 0}, {Mips::FP_NM, 1}, {Mips::RA_NM, 2},  {Mips::S0_NM, 3},
+      {Mips::S1_NM, 4}, {Mips::S2_NM, 5}, {Mips::S3_NM, 6},  {Mips::S4_NM, 7},
+      {Mips::S5_NM, 8}, {Mips::S6_NM, 9}, {Mips::S7_NM, 10},
+  };
+
   // nanoMIPS save and restore instructions require callee-saved registers to be
   // saved in particular order on the stack.
-  auto SortCalleeSaves = [](CalleeSavedInfo First, CalleeSavedInfo Second) {
-    std::unordered_map<unsigned, unsigned> Regs{
-        {Mips::GP_NM, 0}, {Mips::FP_NM, 1}, {Mips::RA_NM, 2},  {Mips::S0_NM, 3},
-        {Mips::S1_NM, 4}, {Mips::S2_NM, 5}, {Mips::S3_NM, 6},  {Mips::S4_NM, 7},
-        {Mips::S5_NM, 8}, {Mips::S6_NM, 9}, {Mips::S7_NM, 10},
-    };
-
+  auto SortCalleeSaves = [&Regs](CalleeSavedInfo First, CalleeSavedInfo Second) {
     // There should be no callee-saved registers that are not part of the list.
     assert(Regs.find(First.getReg()) != Regs.end() &&
            Regs.find(Second.getReg()) != Regs.end());
 
     return Regs[First.getReg()] < Regs[Second.getReg()];
   };
+
+  // If CSI list has less than two callee-saved registers we can
+  // return from method since no insertions nor sorting is needed
+  if(CSI.size() < 2)
+    return false;
+
+  auto MaxCS = std::max_element(
+      CSI.begin(), CSI.end(), [&Regs](CalleeSavedInfo &a, CalleeSavedInfo &b) {
+        return Regs.at(a.getReg()) < Regs.at(b.getReg());
+      });
+  auto MaxCSNum = Regs.at(MaxCS->getReg());
+
+  auto MinCS = std::min_element(
+      CSI.begin(), CSI.end(), [&Regs](CalleeSavedInfo &a, CalleeSavedInfo &b) {
+        return Regs.at(a.getReg()) < Regs.at(b.getReg());
+      });
+  auto MinCSNum = Regs.at(MinCS->getReg());
+
+  // Inserting all of the missing callee-saved registers between min and max
+  // in order to allow further load-store optimizations
+  for (unsigned i = MinCSNum + 1; i < MaxCSNum; i++) {
+
+    auto SearchResult =
+        std::find_if(CSI.begin(), CSI.end(), [&Regs, &i](CalleeSavedInfo &p) {
```

Comment:
It seems a bit inefficient to do a linear find for each potential register.

How about, instead:
  - Go through CSI and build a BitSet of CSNums that are in the saves list
  - Iterate over CSNums, any CSNum not in BitSet must be added
  - Find regnum for CSI by `find_if` (this is linear time complexity, but is rare), OR build an inverse mapping `unordered_set` (again, static member) that maps CSNum to register numbe , and use that.

?